### PR TITLE
[IA-4439] Sanitize leo responses types

### PR DESCRIPTION
**Pull request by LizBaldo**
### What
- Optional fields in leonardo's responses were actually not optional, but returning a 'null' this PR is enforcing this and making these fields non optional.

### Why
- We want to avoid potential bugs creeping in from assuming that fields are optional, but they are actually required with a 'null'

### Testing strategy
- [ ]

### CODE REVIEW

**Comment by nawatts** [comment]
I thought the plan was to change the behavior of the serializer in the backend to omit the fields? If we're going to do this in the UI, it seems like it might be simpler to update the types to match what the backend outputs.

**Comment by mlilynolting** [comment]
> I thought the plan was to change the behavior of the serializer in the backend to omit the fields? If we're going to do this in the UI, it seems like it might be simpler to update the types to match what the backend outputs.

@nawatts the thinking here was:
* We prefer the API design of non-null optional fields (value | undefined)
* Changes to the backend response would also affect All of Us, which consumes our API. We don't know what the effect would be without trying it
* Until we do that coordinated change, this is an incremental step to munge our API responses into the shape the frontend prefers. It's then simple to update the API when we are ready

**Comment by mlilynolting** [comment]
The general model I'm conceptualizing is
```
user interface
biz logic
data as state
data normalization <-- this is the layer we are defining
fetch (raw JSON)
---
formatting for client (API)
biz logic
DAO
database
```

**Comment by nawatts** [comment]
Not changing the backend API makes sense. But it seems like we could avoid some complexity by making the frontend prefer the nullable version vs the optional version. That would be more consistent with other Ajax code in Terra UI, which generally does not reshape the data.

One of the things stated in the coding guidelines (though it's in reference to naming things) is:
> The goal is to preserve the data’s fundamental structure as much as possible from get => render => post.
-- https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide#naming-conventions

Preserving the current shape also reduces the risk of breaking any JS code that uses these responses.

**Comment by mlilynolting** [comment]
To your last point -- there are no changes here to the client logic. It already expects the `value | undefined` shape. Here, we are explicitly preparing the data to match the expected contract.

**Comment by mlilynolting** [comment]
This change is iirc an alternative to adding null safes to the general application code

**Comment by mlilynolting** [comment]
I prefer a single explicit point of data stardardization to many inline conditional modifications, such as nullsafes.

**Comment by nawatts** [comment]
> It already expects the value | undefined shape.

_TypeScript_ code does and TS ensures that it can handle that shape. But if there is any JavaScript code using these responses (`git grep` suggests that there is), then it may be relying on the actual current shape of the data.

**Comment by mlilynolting** [comment]
I agree that ideally the API is in sync with the display requirements of the frontend.

**Comment by nawatts** [comment]
> This change is iirc an alternative to adding null safes to the general application code

Is there a substantial difference between needing to be resilient to null vs needing to be resilient to undefined?

**Comment by msilva-broad** [comment]
> I agree that ideally the API is in sync with the display requirements of the frontend.

My rule of thumb is that APIs should have clean contracts that make them good APIs.  It is a separate desire within the UI code to keep api-related data/contract transforms to a minimum.

I am in agreement that Typescript should make "chasing the errors" on any contract change a straightforward exersise.  My vote:

1) remove Sanitized type/conversion pattern and just have the UI types reflect the existing api contract.
2) carefully address any JS usages for correctness and consider at least partial conversion to TS for future type safety and error chasing benefits on contract changes.

**Comment by mlilynolting** [comment]
> > It already expects the value | undefined shape.
> 
> _TypeScript_ code does and TS ensures that it can handle that shape. But if there is any JavaScript code using these responses (`git grep` suggests that there is), then it may be relying on the actual current shape of the data.

Thank you!! @LizBaldo we'll need to update this PR, then.

I agree that if there's lots of legacy JS expecting value | null, changing Typescript's definition would be a smaller change.

I don't think that I agree that our standard preferring consistent data shape means that this kind of normalization step is an antipattern. If we have to do it, we should do it this way (once, at fetch time).

I do see your argument that we don't have to do it, though. Maybe we can loop @jdcanas in on this when he returns from vacation tomorrow.

**Comment by mlilynolting** [comment]
These fields are declared as Scala.Option[T] on the backend. Circe's Encoder converts Option.None => null when marshalling response objects to JSON.

Choosing either `undefined` or `null` to represent an optional-false value and keeping it consistent across terra-ui seems best.

I think the legacy JS usages might have been overlooked, and the Typescript interface definition taken as authoritative, during planning.

**Comment by msilva-broad** [comment]
> These fields are declared as Scala.Option[T] on the backend. Circe's Encoder converts Option.None => null when marshalling response objects to JSON.
> 
> Choosing either `undefined` or `null` to represent an optional-false value and keeping it consistent across terra-ui seems best.
> 
> I think the legacy JS usages might have been overlooked, and the Typescript interface definition taken as authoritative, during planning.

How the JSON shape is formed by the service layer is not as important as the end result... which is missing value == null.  I think we should push for simplicity in the UI code where we can.  Requiring the expected null typing for missing value makes the code simpler  (removes need for sanitization)  so we should refactor to that simpler pattern.

There may be scenarios where a state/data provider or other layer is warranted in the ui stack between the api call (data client) and visual component.  When that added layer is abstracting or adding other value, conversions like this may be warranted... but this feels like a simpler scenario.  So my vote is to refactor for simplicity.

File: src/libs/ajax/leonardo/Runtimes.ts [hunk 1]
```diff
@@ -11,7 +11,12 @@ import {
   jsonBody,
   makeRequestRetry,
 } from 'src/libs/ajax/ajax-common';
-import { GetRuntimeItem, ListRuntimeItem } from 'src/libs/ajax/leonardo/models/runtime-models';
+import {
+  SanitizedGetRuntimeItem,
+  SanitizedListRuntimeItem,
+  sanitizeGetRuntime,
+  sanitizeListRuntime,
```
Comment (jdcanas) [inline src/libs/ajax/leonardo/Runtimes.ts]: The functions and these types should likely live in `Runtimes.ts`, with a type `export type Runtime = SanitizedListRuntimeItem | SanitizedGetRuntimeItem` being the interface outside components program against unless they need to be more specific

Comment (jdcanas) [reply src/libs/ajax/leonardo/Runtimes.ts]: cc @msilva-broad, thoughts? Would you agree with the above?

File: src/libs/ajax/leonardo/models/runtime-models.ts [hunk 2]
```diff
@@ -67,6 +82,9 @@ export interface ListRuntimeItem {
   patchInProgress: boolean;
 }
 
+export const sanitizeListRuntime = (listRuntime: ListRuntimeItem): SanitizedListRuntimeItem =>
+  _.omitBy((value, key) => key === 'workspaceId' && value === null, listRuntime) as SanitizedListRuntimeItem;
```
Comment (jdcanas) [inline src/libs/ajax/leonardo/models/runtime-models.ts]: Hmm... can we omit all null values instead? A small worry is that we could sanitize away fields that are accidentally null, but type safety should protect us from removing fields that we weren't expecting to be null, and if the code tries to access a field our types say shouldn't be null, but is, in an inappropriate way then you will get the same runtime exception regardless of of its omitted or present as null.  

i.e.
`_.omitBy(listRuntime, v => v === null);`

File: src/libs/ajax/leonardo/models/runtime-models.ts [hunk 3]
```diff
@@ -114,7 +132,47 @@ export interface GetRuntimeItem {
...
+  autopauseThreshold: number;
+  defaultClientId: string | null;
+  runtimeImages: LeoRuntimeImage[];
+  scopes: string[];
+  customEnvironmentVariables: Record<string, any>;
+  diskConfig: DiskConfig | null;
+  patchInProgress: boolean;
+}
+
+const optionalFields = new Set([
```
Comment (jdcanas) [inline src/libs/ajax/leonardo/models/runtime-models.ts]: Hmm... I don't think this is necessary again. Should be safe to do `_.omitBy(obj, v => v === null);`
@msilva-broad, thoughts about sanitizing nulls out of this object?

To re-cap, the backend swagger docs specify these fields as optional. Unfortunately, the encoding of this object into JSON includes optional fields as `null` in the payload. Here, we bridge the gap between the backend's definition of optional to be in line with typescripts definition of optional by dropping `null` values from the object.

File: src/libs/ajax/leonardo/Runtimes.ts (hunk 1 above)
Comment (jdcanas) [reply src/libs/ajax/leonardo/Runtimes.ts]: Ah, sorry, ignore this, trying to speed through my vacation backlog and totally missed the prior conversation 😬

File: src/libs/ajax/leonardo/models/runtime-models.ts (hunk 3 above)
Comment (jdcanas) [reply src/libs/ajax/leonardo/models/runtime-models.ts]: ignore once again

**Comment by jdcanas** [comment]
> > These fields are declared as Scala.Option[T] on the backend. Circe's Encoder converts Option.None => null when marshalling response objects to JSON.
> > Choosing either `undefined` or `null` to represent an optional-false value and keeping it consistent across terra-ui seems best.
> > I think the legacy JS usages might have been overlooked, and the Typescript interface definition taken as authoritative, during planning.
> 
> How the JSON shape is formed by the service layer is not as important as the end result... which is missing value == null. I think we should push for simplicity in the UI code where we can. Requiring the expected null typing for missing value makes the code simpler (removes need for sanitization) so we should refactor to that simpler pattern.
> 
> There may be scenarios where a state/data provider or other layer is warranted in the ui stack between the api call (data client) and visual component. When that added layer is abstracting or adding other value, conversions like this may be warranted... but this feels like a simpler scenario. So my vote is to refactor for simplicity.

The other reason why we were thinking of using this pattern is that there are several cases where, in addition to sanitizing, we want to decorate the data with certain UI-specific things (for example, the display label we use for a certain field so we don't have to call the same method on the same data everywhere it is displayed). In this case (i.e. for disks) I would vote for using the pattern here, but for the simple case here where we are just cleaning, I can see deferring this pattern until there is decorating of data to be done.

**Comment by nawatts** [comment]
Commenting quickly... in the context of also enriching the responses from Leo, I think this makes more sense.

I think I'd still suggest having Ajax methods return the unmodified response for alignment with other Ajax methods and to avoid the risk that JS code relies on the current shape.

And then adding another layer on top of the low-level Ajax to return the enriched model objects, along the lines of what's described in the [cross-team code sharing proposal](https://docs.google.com/document/d/1vefkF3tbIPgn_2Kbx6lscFK6nYoZoaowlkXXkszxijA/edit#heading=h.cmco44cpjoai).

File: src/libs/ajax/leonardo/models/runtime-models.ts (hunk 2 above)
Comment (nawatts) [reply src/libs/ajax/leonardo/models/runtime-models.ts]: It would be nice to have stronger type safety in this method instead of relying on a cast.

**Comment by msilva-broad** [comment]
> Commenting quickly... in the context of also enriching the responses from Leo, I think this makes more sense.
> 
> I think I'd still suggest having Ajax methods return the unmodified response for alignment with other Ajax methods and to avoid the risk that JS code relies on the current shape.
> 
> And then adding another layer on top of the low-level Ajax to return the enriched model objects, along the lines of what's described in the [cross-team code sharing proposal](https://docs.google.com/document/d/1vefkF3tbIPgn_2Kbx6lscFK6nYoZoaowlkXXkszxijA/edit#heading=h.cmco44cpjoai).

yes, this.

We should avoid adding concerns to things.... the object should not mutate as it travels through the code because additional concerns are being added.  The layer approach was my first thought as well.  And follow immutability principle.  With spread operator it is easy to copy an existing object to a new one that you then also add an extra property to.... or use a parent object that contains the original alongside other info.  This feels like a good candidate for either a data-provider or a state-provider.

Also, wherever things land.... code in the other direction that may be stripping off added things should instead copy over only KNOWN properties which avoids burdening that code with knowledge of other parts of the codebase.

File: src/libs/ajax/leonardo/models/runtime-models.ts (hunk 2 above)
Comment (msilva-broad) [reply src/libs/ajax/leonardo/models/runtime-models.ts]: agreed Nick, I would copy known properly only, which is a bit more code to type out, but:
- is safer
- removes the need for magic knowledge of things in other areas of the code,
- avoid overly general stripping code (later they may be legit properties with value null)
- and should avoid needing to "as MyType" cast at all.

**Comment by LizBaldo** [comment]
So I am still very new to the UI and I have to admit that a lot of these concepts are going a bit over my head. I reverted the sanitization and explicitly made the optional fields required with `something | null`, however, the rest of the code used to be expecting optional fields and now this assumption is broken and is causing tests to fail.

It would make more sense to me to sanitize first and then not modify the rest of the UI code because now the assumption that the fields are optional will be true. Am I wrong? Should I now go through the whole codebase and make sure that we can handle `null`? This would be an anti-pattern for me, but again, i am new to this

**Comment by nawatts** [comment]
The thing is TypeScript can only tell us where our assumptions are wrong in TypeScript code. Changing the types to accurately reflect API responses, TS can tell us exactly what is making invalid assumptions. But if we change the shape of the data by sanitizing it, there are no such automated checks for plain JS code. There may be unit tests for some, but unit test coverage is also concentrated on TS files.

**Comment by nawatts** [comment]
Put another way... currently, TS and JS code have different assumptions about the shape of this data. Currently, JS's assumptions are correct and TS's are incorrect because the types are not accurate.

If we correct the types, TS will helpfully tell us what needs to be fixed, revealing bugs that currently exist.

If we sanitize the responses, we make TS's assumptions correct and invalidate JS's. And it's more difficult to update JS code after that because there's no compiler to point out errors.

File: src/libs/ajax/leonardo/models/runtime-models.ts [hunk 4]
```diff
@@ -55,7 +55,7 @@ export interface RuntimeError extends LeoError {
 
 export interface ListRuntimeItem {
   id: number;
-  workspaceId?: string;
+  workspaceId: string | null;
```
Comment (LizBaldo) [inline src/libs/ajax/leonardo/models/runtime-models.ts:58]: I am not so sure that requiring workspaceId for `ListRuntimeItem`, but not requiring it for `GetRuntimeItem` works now. 

Based on the error message I am seeing when trying to build:
```
Property 'workspaceId' is missing in type '{ status: LeoRuntimeStatus; id: number; runtimeName: string; googleProject: string; cloudContext: CloudContext; serviceAccount: string; asyncRuntimeFields: AsyncRuntimeFields | null; ... 16 more ...; patchInProgress: boolean; }' but required in type 'ListRuntimeItem'
```

It feels to me that the Runtime = ListRuntimeItem | GetRuntimeItem is wrong now

Comment (mlilynolting) [reply src/libs/ajax/leonardo/models/runtime-models.ts:58]: Where are you getting that error?

**Comment by Qi77Qi** [comment]
can u explain a bit more why we're doing this?

is it that leo's response has these `null` values, and UI today doesn't handle that properly? if yes, I think there should be an easy way to make leo response encoder to not stick `null` in these responses

You can add `.mapJson(_.dropNullValues)` on the encoder that you care about this behavior on leo side

**Comment by msilva-broad** [comment]
> can u explain a bit more why we're doing this?
> 
> is it that leo's response has these `null` values, and UI today doesn't handle that properly? if yes, I think there should be an easy way to make leo response encoder to not stick `null` in these responses
> 
> You can add `.mapJson(_.dropNullValues)` on the encoder that you care about this behavior on leo side

Hi Qi,  I can attempt to summarize (@LizBaldo please check my homework)

1) I agree that it would be cleaner to have the service API fixed but as was pointed out, this would break AoU (I think it was) who is consuming this API also

2) The JS code appears to match the current null behavior of the API, so the safer thing to do is to have the TS also match null

I think it would be cleaner if the service API was versioned with the null removal fix, have AoU migrate when they have time, and have Terra UI use the cleaner API sooner.... But this would come with a larger code surface on the UI since JS code is harder/scarier to update.... Ideally we would convert the JS to TS first (which assumes unit test love also), THEN make this bigger improvement to all the ui code since TS will help us chase the errors with way more confidence.

**Comment by Qi77Qi** [comment]
@msilva-broad  I see...that's helpful...so it's something that came up as part of converting to TS? which is why it wasn't an issue before becuz JS allows it?

cc @dmohs as well (see Miguel's comment)

**Comment by nawatts** [comment]
> so it's something that came up as part of converting to TS?

Yes. Since the types for Leo responses didn't match the actual shape of those responses, TypeScript code written using them was built on invalid assumptions.

File: src/libs/ajax/leonardo/models/runtime-models.ts (hunk 4 above)
Comment (mlilynolting) [reply src/libs/ajax/leonardo/models/runtime-models.ts:58]: I pushed a commit which addresses this and gets tests passing - https://github.com/DataBiosphere/terra-ui/commit/edc5910. Passes the mock Runtime through a helper which converts to ListRuntimeItem using an explicitly-defined type guard.

File: src/libs/ajax/leonardo/models/runtime-models.ts (hunk 2 above)
Comment (mlilynolting) [reply src/libs/ajax/leonardo/models/runtime-models.ts]: @msilva-broad what's your preferred approach for copying known properties only?

Comment (nawatts) [reply src/libs/ajax/leonardo/models/runtime-models.ts]: It's verbose, but something like this provides type safety.
```ts
const convertType(obj: InputType): OutputType => {
  return {
    field1: obj.field1,
    field2: obj.field2,
    ...
  };
}
```

**Comment by mlilynolting** [comment]
@msilva-broad and @nawatts, I'm trying to get clear on this. Am I understanding correctly if I summarize--

1. Now that our types (interfaces) are defined as expecting `value | null`, the JS code works as is. The Typescript files have been expecting `value | undefined` so those usages may need to change in this PR.
2. Sanitize early, leave the data shape unchanged downstream throughout the UI. You're proposing a slightly different order of operations, to match [the code sharing proposal](https://docs.google.com/document/d/1vefkF3tbIPgn_2Kbx6lscFK6nYoZoaowlkXXkszxijA/edit#heading=h.cmco44cpjoai)'s component diagram:

`Component (view) <- State provider (hook) <- Provider (sanitizer) <- Data client (Ajax calls)`

The code we're adding here would go in the Provider. So I suppose the revised ask from Miguel and Nick might be:

* Leave the ajax/../Runtimes.ts code alone. The methods return ListRuntimeItem or GetRuntimeItem.
* Add a Provider (would take a `signal`, an action, and arguments, and pass them through to a Runtimes.ts method. Would return a sanitized Runtime type). What would this provider be called? Is there a specific Reacty or Typescripty code structure that is appropriate here?
* Change all calls to methods on Runtimes.ts - in `useRuntimes.ts`, for example, and `WorkspaceContainer.js` - to make a call to the RuntimeProvider instead, and expect Runtime return types.

?

File: src/libs/ajax/leonardo/models/runtime-models.ts (hunk 2 above)
Comment (mlilynolting) [reply src/libs/ajax/leonardo/models/runtime-models.ts]: Got it! That makes sense.

**Comment by nawatts** [comment]
The incorrect types meant that TypeScript code was built on invalid assumptions. Those types have no effect on JS code. Without other information, my assumption has been that the JS code is correct and my claim is that it's easier to update the types and TypeScript code than change the shape of the data and potentially break the JS code.

Most Ajax methods do very little... they add some auth headers, make an API request, and return the data unmodified. Generally, I think we would like to keep with that convention and, instead of expanding the responsibilities of the Ajax layer, add another layer that reshapes the response into a model used by the app.

I don't have particularly strong opinions about the details of that layer's API. I was imagining something along the lines of:
- Ajax.listAppsV2 returns ListAppsResponse
- AppsClient.listApps returns AppModel[]

For the integration with React, the pattern I'd suggest is a custom hook that handles fetching the data and storing it in component state. So something like `useApps(workspaceId)` returns `{ loading, apps, error }`.

**Comment by mlilynolting** [comment]
> The incorrect types meant that TypeScript code was built on invalid assumptions. Those types have no effect on JS code. Without other information, my assumption has been that the JS code is correct and my claim is that it's easier to update the types and TypeScript code than change the shape of the data and potentially break the JS code.
> 
> Most Ajax methods do very little... they add some auth headers, make an API request, and return the data unmodified. Generally, I think we would like to keep with that convention and, instead of expanding the responsibilities of the Ajax layer, add another layer that reshapes the response into a model used by the app.
> 
> I don't have particularly strong opinions about the details of that layer's API. I was imagining something along the lines of:
> 
>     * Ajax.listAppsV2 returns ListAppsResponse
> 
>     * AppsClient.listApps returns AppModel[]
> 
> 
> For the integration with React, the pattern I'd suggest is a custom hook that handles fetching the data and storing it in component state. So something like `useApps(workspaceId)` returns `{ loading, apps, error }`.

In your example, it seems like `useApps` would be the State Provider, `AppsClient` would be the Provider, and `ajax/../Runtimes` would be the Data Client. Does that sound right?

**Comment by mlilynolting** [comment]
I've been looking at `useRuntimes.ts` but just realized it's never used in the code. All dressed up w/ nowhere to go...

**Comment by nawatts** [comment]
> In your example, it seems like useApps would be the State Provider, AppsClient would be the Provider, and ajax/../Runtimes would be the Data Client. Does that sound right?

Yes, that sounds right.

File: src/analysis/utils/disk-utils.ts [hunk 5]
```diff
@@ -53,7 +53,9 @@ export const getCurrentAttachedDataDisk = (
   app: App,
   appDataDisks: PersistentDisk[]
 ): DecoratedPersistentDisk | undefined => {
-  const currentDisk: PersistentDisk | undefined = _.find({ name: app?.diskName }, appDataDisks);
+  const currentDisk: PersistentDisk | undefined =
+    !!app && app.diskName === null ? undefined : appDataDisks.find(({ name }) => app.diskName === name);
```
Comment (mlilynolting) [inline src/analysis/utils/disk-utils.ts]: I think this actually will fail when `app` is undefined.

Comment (mlilynolting) [reply src/analysis/utils/disk-utils.ts]: might need to be `!app || `

File: src/analysis/utils/disk-utils.ts [hunk 6]
```diff
@@ -53,7 +53,10 @@ export const getCurrentAttachedDataDisk = (
   app: App,
   appDataDisks: PersistentDisk[]
 ): DecoratedPersistentDisk | undefined => {
-  const currentDisk: PersistentDisk | undefined = _.find({ name: app?.diskName }, appDataDisks);
+  const currentDisk: PersistentDisk | undefined =
```
Comment (LizBaldo) [inline src/analysis/utils/disk-utils.ts]: @mlilynolting lmk if this is what you had in mind

Comment (mlilynolting) [reply src/analysis/utils/disk-utils.ts]: I think this is not quite right still. `!` operator evaluates to a boolean - so `!app` and `!!app` will return `false` or `true`. So `!!app === null` will always be false (true | false <> null).

I think the unhappy paths we're guarding against in this method are

1. `app` could be `undefined` => method should return `undefined`
2. `app.diskName` could be `null` => method should return `undefined`
3. `app.diskName` could be `""`, the empty string => this is fine, continue
4. `appDataDisks` could contain no matches for `app.diskName` => return `undefined`

Comment (mlilynolting) [reply src/analysis/utils/disk-utils.ts]: Ignore this if it's old hat for you; here's how `foo?.bar` evaluates given different inital states:

<table>
<tr>
 <td>`foo = undefined`
 <td> `undefined`
<tr>
 <td> `foo = {}`
 <td> `undefined`
<tr>
 <td> `foo = { bar: undefined }`
 <td> `undefined`
<tr>
 <td> `foo = { bar: null }`
 <td> `null`
<tr>
 <td> `foo = { bar: "" }`
 <td> `""`
<tr>
 <td> `foo = { bar: "baz" }`
 <td> `"baz"`
</table> 

Here's what `foo.bar` (without the nullsafe `?.` operator) returns:

<table>
<tr>
 <td>`foo = undefined`
 <td> `Uncaught TypeError: g is undefined`
<tr>
 <td> `foo = {}`
 <td> `undefined`
<tr>
 <td> `foo = { bar: undefined }`
 <td> `undefined`
<tr>
 <td> `foo = { bar: null }`
 <td> `null`
<tr>
 <td> `foo = { bar: "" }`
 <td> `""`
<tr>
 <td> `foo = { bar: "baz" }`
 <td> `"baz"`
</table>

Comment (mlilynolting) [reply src/analysis/utils/disk-utils.ts]: I just realized that this method types `app: App` - I am still a Typescript beginner but I think this means we are guaranteeing that `app` is neither `undefined` or `null`. That would mean that it will be reliably truthy (`!!app === true`). So if you wanted you could probably change this to `app.diskName ? undefined : appDataDisks.find...`

Comment (mlilynolting) [reply src/analysis/utils/disk-utils.ts]: but in any case this code will work as written, approved!

**Review by mlilynolting — APPROVED** [review]

File: src/libs/wds-status.test.ts [hunk 7]
```diff
@@ -126,7 +126,6 @@ describe('useWdsStatus', () => {
       },
       appName: 'wds-6601fdbb-4b53-41da-87b2-81385f4a760e',
       appType: 'WDS',
-      // @ts-expect-error Leo app types do not reflect actual API responses.
```
Comment (jdcanas) [inline src/libs/wds-status.test.ts:129]: hmm.. why is this needed?

Comment (LizBaldo) [reply src/libs/wds-status.test.ts:129]: VScode was telling me that they are unused, but I can put it back!

Comment (nawatts) [reply src/libs/wds-status.test.ts:129]: This comment was originally here because I added a realistic mock API response in this test in anticipation of IA-4439 getting resolved. It's good that this comment is no longer necessary.

**Review by aherbst-broad — COMMENTED** [review]
This looks like a reasonable approach. I see checks are failing -- is this ready for review? Give me a ping if so.

**Comment by LizBaldo** [comment]
Thanks :) I am pairing with @jdcanas on Friday to get this through the door, now that some fields are not optional anymore we need to rethink the mocking of the leo's responses a bit more

**Comment by LizBaldo** [comment]
Alright, this is finally ready for review and hopefully two 👍

File: src/analysis/utils/disk-utils.ts [hunk 8]
```diff
@@ -62,7 +62,8 @@ export const getCurrentAttachedDataDisk = (
...
+    !!app && app.diskName === null ? undefined : appDataDisks.find(({ name }) => app.diskName === name);
```
Comment (nawatts) [inline src/analysis/utils/disk-utils.ts]: Is the `!!app` needed here? This function as typed as requiring that `app` be an `App`, not `App | undefined`.

If `app` can actually be undefined, it looks like that is not handled, since the `find` would attempt to read `app.diskName`.

Comment (LizBaldo) [reply src/analysis/utils/disk-utils.ts]: Arg, I think the original typing in the function is wrong, and the App can be undefined :( Let me take another pass at handling that case

File: src/analysis/utils/disk-utils.ts [hunk 9]
```diff
@@ -59,10 +59,11 @@ export const defaultGceBootDiskSize = 120;
 export const defaultGcePersistentDiskSize = 50;
 export const defaultPersistentDiskType = googlePdTypes.standard;
 export const getCurrentAttachedDataDisk = (
-  app: App,
+  app: App | undefined,
   appDataDisks: PersistentDisk[]
 ): DecoratedPersistentDisk | undefined => {
-  const currentDisk: PersistentDisk | undefined = _.find({ name: app?.diskName }, appDataDisks);
+  const currentDisk: PersistentDisk | undefined =
+    app?.diskName === null ? undefined : appDataDisks.find(({ name }) => app?.diskName === name);
```
Comment (nawatts) [inline src/analysis/utils/disk-utils.ts]: If `app` is undefined, `app?.diskName` will be undefined, not null. That would mean that the `appDataDisks.find(...)` path would get run. That might be fine, if disks are guaranteed to have a name, then looking for one with an undefined name wouldn't return a result. That's what the current behavior is if `app` is undefined.

Alternatively, it might be worth loosening the condition to check if `app?.diskName` is falsy instead of strictly checking for null.

```suggestion
    !app?.diskName ? undefined : appDataDisks.find(({ name }) => app.diskName === name);
```

**Review by nawatts — APPROVED** [review]

File: src/analysis/utils/disk-utils.ts (hunk 9 above)
Comment (LizBaldo) [reply src/analysis/utils/disk-utils.ts]: I see, I am still new to the lodash syntax. Basically an app will always have a diskName, but it might be null, and in that case i want to return `undefined` for the persistent disk, which I think your suggested syntax is doing. Thanks for the explanation and pointers!

Comment (nawatts) [reply src/analysis/utils/disk-utils.ts]: The `?.` in `app?.diskName` is the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) and is part of JavaScript, not Lodash.

`app?.diskName` is effectively `app === undefined || app === null ? undefined : app.diskName`.

**Review by mspector — APPROVED** [review]
